### PR TITLE
Add prototype of cloud functions feature -- email sending part.

### PR DESCRIPTION
**Pull request by LightFelicis**
This PR consists of 2 steps:
* Set up cloud functions,
* Create email sending cloud function.

Sending emails feature is just a prototype, because ideally it would react on
firestore changes. Firestore part is not merged yet, so right now the email sending
uses HTTPS requests on fixed cloud function endpoint.

Authorisation process is the same, so I think it may be valuable to split
prototype and final versions into 2 PRs -- this one and one more after
the firestore part is merged into master :-)

### CODE REVIEW

File: functions/index.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+// time.
+// If modifying the scopes below, delete token.json.
+const SCOPES = [
+  'https://mail.google.com/',
+  'https://www.googleapis.com/auth/gmail.modify',
+  'https://www.googleapis.com/auth/gmail.compose',
+  'https://www.googleapis.com/auth/gmail.send',
+];
+
+exports.sendMail = functions.https.onRequest((request, response) => {
```
Comment (LightFelicis) [inline functions/index.js:20]: This line should be changed in the future -- from httpsRequest to firestore listener.

File: firebase.json [hunk 2]
```diff
@@ -4,11 +4,22 @@
...
+    "rewrites": [
+      {
+        "source": "**",
+        "destination": "/index.html"
+      }
+    ]
+  },
+  "functions": {
+    "predeploy": [
+      "npm --prefix \"$RESOURCE_DIR\" run lint"
```
Comment (leon-kastler-google) [inline firebase.json:22]: the linter setup could have been a different PR ;-)

Comment (LightFelicis) [reply firebase.json:22]: Sorry, it's automatically added by cloud functions init. I will split it :)

Comment (LightFelicis) [reply firebase.json:22]: I've added additional PR only for that files. I'll rebase this branch as soon as the "init cloud functions" is merged into master.

Comment (LightFelicis) [reply firebase.json]: Done.

File: functions/index.js [hunk 3]
```diff
@@ -1,9 +1,104 @@
...
+    });
+  });
+}
+
+function makeBody(to, from, subject, message) {
+  const str = [
+    'Content-Type: text/plain; charset="UTF-8"\n',
+    'MIME-Version: 1.0\n',
+    'Content-Transfer-Encoding: 7bit\n',
+    'to: ', to, '\n',
```
Comment (xmxwx) [inline functions/index.js:75]: nit
```suggestion
    'To: ', to, '\n',
```

File: functions/index.js [hunk 4]
```diff
@@ -1,9 +1,104 @@
...
+    'from: ', from, '\n',
+    'subject: ', subject, '\n\n',
```
Comment (xmxwx) [inline functions/index.js:77]: nit
```suggestion
    'Subject: ', subject, '\n\n',
```

File: functions/index.js [hunk 5]
```diff
@@ -1,9 +1,104 @@
...
+  ].join('');
+
+  const encodedMail = new Buffer.from(str).toString('base64')
+      .replace(/\+/g, '-')
+      .replace(/\//g, '_');
+  return encodedMail;
+}
+
+function sendMessage(auth, recipient) {
+  const raw = makeBody(recipient, 'kluczek@identity-sre.com',
```
Comment (xmxwx) [inline functions/index.js:88]: Idea: you can obtain the sender address from [users.getProfile](https://developers.google.com/gmail/api/reference/rest/v1/users/getProfile).

**Review by xmxwx — APPROVED** [review]

**Review by leon-kastler-google — APPROVED** [review]

File: functions/index.js (hunk 5 above)
Comment (LightFelicis) [reply functions/index.js:88]: I'll create a ticket for that.

File: functions/index.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  'https://www.googleapis.com/auth/gmail.compose',
+  'https://www.googleapis.com/auth/gmail.send',
+];
+
+exports.sendMail = functions.https.onRequest((request, response) => {
+  const recipient = request.query.recipient;
+  // TODO: Respond with more meaningful response.
+  response.send('Hello from Firebase!');
+  fs.readFile('credentials.json', (err, content) => {
+    if (err) return console.log('Error loading client secret file:', err);
```
[the remark this answers is not in the document]
Comment (LightFelicis) [reply functions/index.js:25]: I've changed if statement to version with brackets.

File: functions/index.js (hunk 1 above)
Comment (LightFelicis) [reply functions/index.js:21]: Done.

File: functions/index.js (hunk 4 above)
Comment (LightFelicis) [reply functions/index.js]: Done.

File: functions/index.js (hunk 3 above)
Comment (LightFelicis) [reply functions/index.js]: Done.

**Pull request merged**